src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs

```diff
@@ -129,31 +110,36 @@ void IDisposable.Dispose()
         /// <param name="inMemoryPdbSize">in memory PDB size</param>
         /// <returns>reader</returns>
         /// <remarks>
-        /// Assumes that neither PE image nor PDB loaded into memory can be unloaded or moved around.
+        /// Accounts for unloadable and dynamic types by keying the cache on the managed Assembly. The
+        /// underlying ConditionalWeakTable doesn't keep the assembly alive, so cached types will be
+        /// correctly invalidated when the Assembly is unloaded by the GC.
         /// </remarks>
-        private unsafe MetadataReader TryGetReader(string assemblyPath, IntPtr loadedPeAddress, int loadedPeSize, IntPtr inMemoryPdbAddress, int inMemoryPdbSize)
+        private unsafe MetadataReader TryGetReader(Assembly assembly, string assemblyPath, IntPtr loadedPeAddress, int loadedPeSize, IntPtr inMemoryPdbAddress, int inMemoryPdbSize)
         {
             if ((loadedPeAddress == IntPtr.Zero || assemblyPath == null) && inMemoryPdbAddress == IntPtr.Zero)
             {
                 // Dynamic or in-memory module without symbols (they would be in-memory if they were available).
                 return null;
             }
 
-            IntPtr cacheKey = (inMemoryPdbAddress != IntPtr.Zero) ? inMemoryPdbAddress : loadedPeAddress;
-
             MetadataReaderProvider provider;
-            while (!_metadataCache.TryGetValue(cacheKey, out provider))
+            while (!_metadataCache.TryGetValue(assembly, out provider))
```

Comment:
It seems it would be better to use the GetValue method that has a callback to create value if it doesn't exist. The doc says the operation is atomic.

Reply:
You're right, the comments indicate this method already accounts for multiple threads attempting to create an entry with the same key

https://github.com/dotnet/coreclr/blob/9d85a643eb32d57188ab1805998dc9a9d6978f8a/src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/ConditionalWeakTable.cs#L166-L193